tool/tsh/kube.go

```diff
@@ -570,40 +570,70 @@ func newKubeCredentialsCommand(parent *kingpin.CmdClause) *kubeCredentialsComman
 }
 
 func (c *kubeCredentialsCommand) run(cf *CLIConf) error {
+	// client.LoadKeysToKubeFromStore function is used to speed up the credentials
+	// loading process since Teleport Store transverses the entire store to find the keys.
+	// This operation takes a long time when the store has a lot of keys and when
+	// we call the function multiple times in parallel.
+	// Although client.LoadKeysToKubeFromStore function speeds up the process
+	// since it removes all transversals, it still has to read 4 different files:
+	// - $TSH_HOME/current_profile
+	// - $TSH_HOME/$profile.yaml
+	// - $TSH_HOME/keys/$PROXY/$USER-kube/$TELEPORT_CLUSTER/$KUBE_CLUSTER-x509.pem
+	// - $TSH_HOME/keys/$PROXY/$USER
+	if kubeCert, privKey, err := client.LoadKeysToKubeFromStore(
+		cf.HomePath,
+		c.teleportCluster,
+		c.kubeCluster,
+	); err == nil {
+		crt, _ := tlsca.ParseCertificatePEM(kubeCert)
+		if crt != nil && time.Until(crt.NotAfter) > time.Minute {
+			log.Debugf("Re-using existing TLS cert for Kubernetes cluster %q", c.kubeCluster)
+			return c.writeByteResponse(cf.Stdout(), kubeCert, privKey, crt.NotAfter)
+		}
+	}
+
 	tc, err := makeClient(cf, true)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
+	_, span := tc.Tracer.Start(cf.Context, "tsh/kubeCredentials/GetKey")
 	// Try loading existing keys.
 	k, err := tc.LocalAgent().GetKey(c.teleportCluster, client.WithKubeCerts{})
```

Comment:
nit: IIUC, it's the `TrustedCertsStore` that performs a file directory traversals, causing the performance issue. Maybe we can achieve the same performance improvement by adding a `client.WithoutCertificateAuthorityCerts{}`? Or even just create a `client.NewFSStore` but replace `store.TrustedCertsStore = client.NewMemTrustedCertsStore` as a quick hack.

This would just be useful to avoid a lot of this custom logic and potentially apply this fix elsewhere when needed.

Reply:
We also have transversals here:
https://github.com/gravitational/teleport/blob/4b7ab4257c11f05764ef1cb3e5836edfcd40dc65/lib/client/keystore.go#L300
And here:
https://github.com/gravitational/teleport/blob/4b7ab4257c11f05764ef1cb3e5836edfcd40dc65/lib/client/keystore.go#L343

Although we don't read on all of them, the operation is slow

Reply:
The first one isn't a traversal, it just reads the directory to see if it's there.

The second *can* be a traversal, but only if `teleportCluster == ""`. I believe the call used (below) would not result in a traversal here.
```
k, err := tc.LocalAgent().GetKey(c.teleportCluster, client.WithKubeCerts{})
```

Reply:
Nevermind, it looks like `client.WithKubeCerts` always leads to a traversal.

Reply:
You could add a `client.WithKubeCert{ KubeClusterName }` to read the single file directly instead.

Reply:
Sure, but when the storage is initialized we also read secrets in `makeClientForProxy` https://github.com/gravitational/teleport/blob/52351314fa88c68647bb96509bfb10684d70c581/tool/tsh/tsh.go#L3427

And for kube credentials, we don't need them at all.
In a normal initialization with a single cluster, tsh ends up reading:

`initProxyClient`
- `/home/user/.tsh/current_profile`
- `/home/user/.tsh/tele.local.yaml`
- `/home/user/.tsh/keys/tele.local/user-x509.pem`
- `/home/user/.tsh/keys/tele.local/user`
- `/home/user/.tsh/keys/tele.local/user.pub`
- `/home/user/.tsh/keys/tele.local/user-ssh/tele.local-cert.pub`

While reading the kube keys:
- `/home/user/.tsh/keys/tele.local/user-x509.pem`
- `/home/user/.tsh/keys/tele.local/user`
- `/home/user/.tsh/keys/tele.local/user.pub`
- `/home/user/.tsh/keys/tele.local/user-kube/tele.local/local-x509.pem`


It reads multiple times the same files and does `os.ReadDir` + transversal reads.

To fix it. I would need to add a cached version to avoid re-reading the same files and introduce the `client.WithKubeCert{ KubeClusterName }` option